enable traces (sub/child-errors) and custom "error types"

oli-obk opened the pull request:
This adds `warningMatch` and improves `functionMatch` to allow build providers to emit almost any kind of lint that is specified here: https://github.com/steelbrain/linter/wiki/Linter-API#messages

An example of a modified `build-cargo`:
![pic of traces and custom error types](https://cloud.githubusercontent.com/assets/332036/15097688/ddfc170c-1523-11e6-8394-d24a79d125ea.png)

fixes #408

noseglid commented on lib/error-matcher.js:
```diff
@@ -68,18 +68,37 @@ export default class ErrorMatcher extends EventEmitter {
-        require('xregexp').forEach(this.output, regex, matchFunction.bind(this.currentMatch));
+
+    // first run all functional matches
+    this.functions && this.functions.forEach((f, functionIndex) => {
```
since `this.functions` is set in the `set` function I feel we can safely assume this will always be an array

oli-obk replied:
javascript scares me xD I add checks everywhere. But yea, I'll remove this check

oli-obk replied:
ah, no it's not always set, only if `functionMatch` is a value.

noseglid commented on lib/build.js:
```diff
@@ -180,22 +180,53 @@ export default {
-            [ (match.line_end || match.line || 1) - 1, (match.col_end || match.col || 1) - 1 ]
-          ]
+          filePath: normalizePath(match.file),
+          severity: cleanSeverity(match.severity) || typeToSeverity(match.type) || 'error',
```
I feel this may be a bit too much. Why do we need to introduce both `type` and `severity` ?

oli-obk replied:
the `Explanation` example uses `Explanation` as the `type` and `info` as the severity. Alternatively we can treat everything that's not `Error`, `Err`, `Warn` or `Warning` as `info`

noseglid commented on lib/build.js:
```diff
@@ -180,22 +180,53 @@ export default {
+            case 'warning': return 'warning';
+            case 'trace': return 'info';
+            case 'note': return 'info';
+            default: return null;
```
Hmm, I have concerns about this. where did `trace` and `note` come from ? It doesn't seem to have anything to do with the generic `build` tool. SInce linter supports `error`, `warning` and `info` I think we should keep those levels too.

But this comment would be unnecessary if we didn't have both `type` and `severity`.

noseglid commented:
This amount of code almost warrants moving all linter related code to a `linter-integration.js` file.

oli-obk commented:
I removed the criticality, and am parsing the type now. I added a few possible type values. Matches with unknown type are turned into `error`, Children of matches with unknown type are turned into `info`.

oli-obk commented:
> This amount of code almost warrants moving all linter related code to a linter-integration.js file.

I tried that, I failed, not sure what I did wrong.

oli-obk commented:
nevermind, I found it :) works now

noseglid commented:
Shaping up nicely! I'll give this a test and a final review later today!

noseglid commented on lib/error-matcher.js:
```diff
@@ -68,42 +68,61 @@ export default class ErrorMatcher extends EventEmitter {
+    }
+    this.regex = {
+      Error: this._prepareRegex(target.errorMatch),
+      Warning: this._prepareRegex(target.warningMatch)
```
Any reason why the keys here are capitalized?

oli-obk replied:
atom-build currently prints `Error` capitalized, I'm generating that name from the field name.

noseglid replied:
Allright, makes sense!

noseglid commented:
Functionalitywise this is great, but I think we need some more documentation around this, with examples on how to add trace etc. The image you have in this PR would also be nice to spark interest!

oli-obk commented on lib/error-matcher.js:
```diff
@@ -68,42 +68,61 @@ export default class ErrorMatcher extends EventEmitter {
+      this.regex[kind] && this.regex[kind].forEach((regex, i) => {
+        regex && require('xregexp').forEach(this.output, regex, (match, matchIndex) => {
+          match.id = 'error-match-' + i + '-' + matchIndex;
+          match.type = kind;
```
here I use the field name as the type

oli-obk commented:
I'll think of an introductory example, but I'm not much of a javascript programmer, so my examples might be rather limited.

noseglid commented:
I think one of the specs should do it. Not much is required so one can get started.

oli-obk commented:
I added a new example and a spec testing the example

noseglid commented:
I'm totally happy with the code and docs now.
Something broke on Windows though :(

oli-obk commented:
probably the `/` path separators, I'll investigate

noseglid commented on spec/build-error-match-spec.js:
```diff
@@ -577,6 +611,68 @@ describe('Error Match', () => {
+      fs.writeFileSync(directory + 'change_dir_output.txt', fs.readFileSync(__dirname + '/fixture/change_dir_output.txt'));
+      fs.mkdirSync(directory + 'foo');
+      fs.mkdirSync(directory + 'foo/src');
+      fs.writeFileSync(directory + 'foo/src/testmake.c', 'lorem ipsum\naquarium laudanum\nbabaorum petibonum\nthe cake is a lie');
```
THis could be the issue. Should probably use `path.join`.

oli-obk commented:
atom still keeps a handle on a file in the directory while we try to delete the directory after each spec run. This issue is talked about here: https://github.com/isaacs/rimraf/issues/72

Not sure how to fix this, because I have no idea how to close a tab in code. And even then, atom might keep the handle longer than it should.

Can we just stop deleting the specs after running them? They are in a temp directory with a unique id anyway.

noseglid commented:
But this only happend after your last change. Previously it was building fine!

oli-obk commented:
I'm actually creating subdirectories. Atom might somehow lock them.

noseglid commented:
The stack trace is originating from `https://github.com/noseglid/atom-build/pull/418/files#diff-bd51e8ec703453db2d933b31b969ab0aR58`.
I'd suggest putting a try-catch around that, if we fail to clean up, buhoo. It'll all work anyway and the next tests gets a new temporary directory anyway.

oli-obk commented:
> I'd suggest putting a try-catch around that, if we fail to clean up, buhoo. It'll all work anyway and the next tests gets a new temporary directory anyway.

That's what I meant, for simplicity we can just not clean up at all :) It's not like we're producing huge files during the specs.

noseglid commented:
I'd feel better with a try-catch than completely removing it though. When I do development I could potentially end up with a lot of temporary files which could've been cleaned up

noseglid commented:
Just put a comment in the catch
```
catch (err) {
  // Failed to clean up, ignore this.
}
```

run `npm test` locally to lint check it.

oli-obk commented:
I edited it throug the github interface b/c I wasn't on a pc that I had set up

noseglid commented:
Merged! Thanks so much for this and all your patience!
